### PR TITLE
executor, privileges: fix privileges check fail for `SET DEFAULT ROLE`

### LINKED ISSUE #11165: `SET DEFAULT ROLE` fails privilege check
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.
```
create user u1@localhost;
create role r1;
grant r1 to u1@localhost;

login as u1:
set default role all to u1@localhost;
```

2. What did you expect to see?
```
mysql> set default role all to u1@localhost;
Query OK, 0 rows affected (0.01 sec)
```


3. What did you see instead?
```
mysql> set default role all to u1@localhost;
ERROR 1227 (42000): Access denied; you need (at least one of) the CREATE USER privilege(s) for this operation
```


4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?
```
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                                                                     |
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v3.0.0-rc.1-273-gf2841dec9
Git Commit Hash: f2841dec95bf81158aa55a0e26aecb318a3dc208
Git Branch: master
UTC Build Time: 2019-07-10 05:57:14
GoVersion: go version go1.12 darwin/amd64
Race Enabled: false
TiKV Min Version: 2.1.0-alpha.1-ff3dd160846b7d1aed9079c389fc188f7f5ea13e
Check Table Before Drop: false |
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

```

### PR DESCRIPTION
**Pull request by imtbkcat**
### What problem does this PR solve? 
fix #11165 
`SET DEFAULT ROLE` should always be avaible for `current_user`

### What is changed and how it works?
Implement `setDefaultRoleForCurrentUser`.
Remove original privilege check for `SET DEFAULT ROLE`.
Add new privilege check for `SET DEFAULT ROLE`.
If is setting default role for current user, just skip check and call `setDefaultRoleForCurrentUser`.

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

Side effects

 - Increased code complexity

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

**Comment by jackysp** [comment]
PTAL @tiancaiamao @lysu

File: executor/simple.go [hunk 1]
```diff
@@ -230,7 +231,64 @@ func (e *SimpleExec) setDefaultRoleAll(s *ast.SetDefaultRoleStmt) error {
...
+				sql += ","
+			}
+		}
+	}
+	deleteSQL := fmt.Sprintf("DELETE IGNORE FROM mysql.default_roles WHERE USER='%s' AND HOST='%s';", user.Username, user.Hostname)
+	_, _, err = sqlExecutor.ExecRestrictedSQL(e.ctx, deleteSQL)
+	if err != nil {
+		return
+	}
+	_, _, err = sqlExecutor.ExecRestrictedSQL(e.ctx, sql)
```
Comment (tiancaiamao) [inline executor/simple.go]: If step1 success and step2 fail... should wrap the operations in a transaction?

Comment (imtbkcat) [reply executor/simple.go]: @tiancaiamao I'm looking for an interface which can operate sql in a transaction without privilege check.

Comment (tiancaiamao) [reply executor/simple.go]: It seems that all sessions from the session pool skip the privilege check.

Comment (imtbkcat) [reply executor/simple.go]: @tiancaiamao sessions in session pool is auto commit.

Comment (imtbkcat) [reply executor/simple.go]: @tiancaiamao I use system session to run sql in transaction, PTAL.

File: executor/simple.go [hunk 2]
```diff
@@ -230,7 +251,89 @@ func (e *SimpleExec) setDefaultRoleAll(s *ast.SetDefaultRoleStmt) error {
...
+				return ErrRoleNotGranted.GenWithStackByArgs(role.String(), user.String())
+			}
+			sql += fmt.Sprintf("('%s', '%s', '%s', '%s')", user.Hostname, user.Username, role.Hostname, role.Username)
+			if i != len(s.RoleList)-1 {
+				sql += ","
+			}
+		}
+	}
+
+	restrictedCtx, err := e.getSysSession()
```
Comment (tiancaiamao) [inline executor/simple.go]: Why not use "begin ... commit/rollback"?

File: executor/simple.go [hunk 3]
```diff
@@ -51,6 +53,24 @@ type SimpleExec struct {
...
+	}
+	restrictedCtx := ctx.(sessionctx.Context)
+	restrictedCtx.GetSessionVars().InRestrictedSQL = true
+	return restrictedCtx, nil
+}
+
+func (e *SimpleExec) releaseSysSession(ctx sessionctx.Context) {
+	dom := domain.GetDomain(e.ctx)
+	sysSessionPool := dom.SysSessionPool()
+	sysSessionPool.Put(ctx.(pools.Resource))
```
Comment (tiancaiamao) [inline executor/simple.go]: `sysSessionPool.Put(ctx)` should work, and import "github.com/ngaut/pools" is not necessary.

**Comment by tiancaiamao** [comment]
LGTM @lysu

**Pull request merged**